File: python/s3_uploader/core/task_runner.py
"""アップロードタスクの実行"""
import os
from typing import List, Tuple

from ..models.config import UploadTask, Config
from ..utils.logger import LoggerManager
from ..utils.file_utils import FileScanner
from .uploader import UploadExecutor, ParallelUploadExecutor
from .s3_client import S3ClientManager


class TaskRunner:
    """アップロードタスクを実行"""
    
    def __init__(self, config: Config):
        self.config = config
        self.logger = LoggerManager.get_logger()
        
        # S3クライアントとアップローダーを初期化
        client_manager = S3ClientManager(config.aws)
        self.s3_client = client_manager.get_client()
        self.executor = UploadExecutor(self.s3_client, config.options)
        self.parallel_executor = ParallelUploadExecutor(
            self.executor, 
            config.options.parallel_uploads
        )
        self.file_scanner = FileScanner(config.options.exclude_patterns)
        
    def run_all_tasks(self) -> Tuple[int, int]:
        """全てのタスクを実行"""
        total_tasks = len(self.config.upload_tasks)
        successful_tasks = 0
        failed_tasks = 0
        
        self.logger.info(f"Starting upload tasks: {total_tasks} tasks to process")
        
        for i, task in enumerate(self.config.upload_tasks, 1):
            if not task.enabled:
                self.logger.info(f"Skipping disabled task: {task.name}")
                continue
                
            self.logger.info(f"Task {i}/{total_tasks}: Starting '{task.name}'")
            
            try:
                success = self._run_single_task(task)
                if success:
                    successful_tasks += 1
                    self.logger.info(f"Task {i}/{total_tasks}: '{task.name}' completed successfully")
                else:
                    failed_tasks += 1
                    self.logger.error(f"Task {i}/{total_tasks}: '{task.name}' failed")
            except Exception as e:
                failed_tasks += 1
                self.logger.error(f"Task {i}/{total_tasks}: '{task.name}' failed with error: {e}")
                
        self.logger.info(
            f"Upload tasks completed: {successful_tasks} successful, {failed_tasks} failed"
        )
        return successful_tasks, failed_tasks
        
    def _run_single_task(self, task: UploadTask) -> bool:
        """単一タスクを実行"""
        if os.path.isfile(task.source):
            # 単一ファイルのアップロード
            return self._upload_single_file(task)
        elif os.path.isdir(task.source):
            # ディレクトリのアップロード
            return self._upload_directory(task)
        else:
            self.logger.error(f"Source is neither file nor directory: {task.source}")
            return False
            
    def _upload_single_file(self, task: UploadTask) -> bool:
        """単一ファイルをアップロード"""
        if not task.s3_key:
            self.logger.error(f"s3_key is required for file upload: {task.name}")
            return False
            
        try:
            file_info = self.file_scanner.get_file_info(task.source)
            result = self.executor.upload_file(file_info, task.bucket, task.s3_key)
            return result.success
        except Exception as e:
            self.logger.error(f"Error uploading file {task.source}: {e}")
            return False
            
    def _upload_directory(self, task: UploadTask) -> bool:
        """ディレクトリをアップロード"""
        s3_key_prefix = task.s3_key_prefix or ""
        
        try:
            # アップロードタスクを収集
            upload_tasks = []
            for file_info in self.file_scanner.scan_directory(task.source, task.recursive):
                s3_key = s3_key_prefix + file_info.relative_path.replace(os.sep, "/")
                upload_tasks.append((file_info, task.bucket, s3_key))
                
            if not upload_tasks:
                self.logger.warning(f"No files found in {task.source}")
                return True
                
            # 並列アップロード実行
            successful, failed = self.parallel_executor.upload_files(upload_tasks)
            
            self.logger.info(
                f"Directory upload completed: {successful} successful, {failed} failed"
            )
            return failed == 0
            
        except Exception as e:
            self.logger.error(f"Error uploading directory {task.source}: {e}")
            return False

File: go/internal/uploader/task_runner.go
package uploader

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"s3-uploader/internal/aws"
	"s3-uploader/internal/logger"
	"s3-uploader/internal/models"
)

// TaskRunner タスクの実行を管理する構造体
type TaskRunner struct {
	uploader *Uploader
	config   models.Config
	logger   *logger.Logger
}

// NewTaskRunner 新しいTaskRunnerを作成
func NewTaskRunner(client aws.S3Operations, config models.Config) *TaskRunner {
	lgr := logger.GetLogger()
	uploader := NewUploader(client, config.Options)
	
	return &TaskRunner{
		uploader: uploader,
		config:   config,
		logger:   lgr,
	}
}

// TaskResult タスク実行結果
type TaskResult struct {
	TaskName      string
	StartTime     time.Time
	EndTime       time.Time
	Duration      time.Duration
	TotalFiles    int
	SuccessFiles  int
	FailedFiles   int
	SkippedFiles  int
	TotalBytes    int64
	UploadResults []UploadResult
	Error         error
}

// RunReport 実行レポート
type RunReport struct {
	StartTime    time.Time
	EndTime      time.Time
	Duration     time.Duration
	TotalTasks   int
	SuccessTasks int
	FailedTasks  int
	SkippedTasks int
	TaskResults  []TaskResult
	DryRun       bool
}

// RunAllTasks すべてのタスクを実行
func (tr *TaskRunner) RunAllTasks(ctx context.Context) (*RunReport, error) {
	report := &RunReport{
		StartTime:   time.Now(),
		DryRun:      tr.config.Options.DryRun,
		TaskResults: make([]TaskResult, 0),
	}

	if tr.config.Options.DryRun {
		tr.logger.Info("🏃 Running in DRY RUN mode - no files will be uploaded")
	}

	tr.logger.Info("Starting task runner",
		"total_tasks", len(tr.config.UploadTasks),
		"dry_run", tr.config.Options.DryRun,
	)

	// 各タスクを実行
	for _, task := range tr.config.UploadTasks {
		// タスクが無効の場合はスキップ
		if !task.Enabled {
			tr.logger.Info("Skipping disabled task", "task", task.Name)
			report.SkippedTasks++
			continue
		}

		tr.logger.Info("📋 Starting task",
			"name", task.Name,
			"description", task.Description,
		)

		result := tr.runTask(ctx, task)
		report.TaskResults = append(report.TaskResults, result)

		if result.Error != nil {
			report.FailedTasks++
			tr.logger.Error("❌ Task failed",
				"task", task.Name,
				"error", result.Error,
			)
		} else {
			report.SuccessTasks++
			tr.logger.Info("✅ Task completed",
				"task", task.Name,
				"duration", result.Duration,
				"files", result.SuccessFiles,
				"bytes", result.TotalBytes,
			)
		}
	}

	report.EndTime = time.Now()
	report.Duration = report.EndTime.Sub(report.StartTime)
	report.TotalTasks = len(tr.config.UploadTasks)

	return report, nil
}

// runTask 個別のタスクを実行
func (tr *TaskRunner) runTask(ctx context.Context, task models.UploadTask) TaskResult {
	result := TaskResult{
		TaskName:  task.Name,
		StartTime: time.Now(),
	}

	// ソースの存在確認
	sourceInfo, err := os.Stat(task.Source)
	if err != nil {
		result.Error = fmt.Errorf("source not found: %w", err)
		result.EndTime = time.Now()
		result.Duration = result.EndTime.Sub(result.StartTime)
		return result
	}

	// ディレクトリかファイルかで処理を分岐
	if sourceInfo.IsDir() {
		// ディレクトリのアップロード
		recursive := task.Recursive // Recursiveは直接bool型

		// S3キープレフィックスを決定
		keyPrefix := ""
		if task.S3KeyPrefix != nil {
			keyPrefix = *task.S3KeyPrefix
		}

		tr.logger.Debug("Uploading directory",
			"source", task.Source,
			"bucket", task.Bucket,
			"prefix", keyPrefix,
			"recursive", recursive,
		)

		// リトライ機能を使ったディレクトリアップロード
		uploadResults, err := tr.uploader.UploadDirectoryWithRetry(ctx, task.Source, task.Bucket, keyPrefix, recursive)
		result.UploadResults = uploadResults
		result.Error = err

		// 結果を集計
		for _, ur := range uploadResults {
			result.TotalFiles++
			if ur.Success {
				if ur.SkippedReason != "" {
					result.SkippedFiles++
				} else {
					result.SuccessFiles++
					result.TotalBytes += ur.Size
				}
			} else {
				result.FailedFiles++
			}
		}
	} else {
		// 単一ファイルのアップロード
		var key string
		if task.S3Key != nil && *task.S3Key != "" {
			key = *task.S3Key
		} else {
			// S3キーが指定されていない場合はファイル名を使用
			key = filepath.Base(task.Source)
		}

		tr.logger.Debug("Uploading file",
			"source", task.Source,
			"bucket", task.Bucket,
			"key", key,
		)

		// リトライ機能を使ったファイルアップロード
		uploadResult, err := tr.uploader.UploadFileWithRetry(ctx, task.Source, task.Bucket, key)
		result.UploadResults = []UploadResult{*uploadResult}
		result.Error = err
		result.TotalFiles = 1

		if uploadResult.Success {
			if uploadResult.SkippedReason != "" {
				result.SkippedFiles = 1
			} else {
				result.SuccessFiles = 1
				result.TotalBytes = uploadResult.Size
			}
		} else {
			result.FailedFiles = 1
		}
	}

	result.EndTime = time.Now()
	result.Duration = result.EndTime.Sub(result.StartTime)

	return result
}

// PrintReport レポートをコンソールに出力
func (tr *TaskRunner) PrintReport(report *RunReport) {
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("📊 UPLOAD TASK RUNNER REPORT")
	fmt.Println(strings.Repeat("=", 60))

	if report.DryRun {
		fmt.Println("🏃 Mode: DRY RUN (no files were actually uploaded)")
	} else {
		fmt.Println("🏃 Mode: LIVE")
	}

	fmt.Printf("⏱️  Total Duration: %s\n", report.Duration.Round(time.Millisecond))
	fmt.Printf("📅 Start Time: %s\n", report.StartTime.Format("2006-01-02 15:04:05"))
	fmt.Printf("📅 End Time: %s\n", report.EndTime.Format("2006-01-02 15:04:05"))
	fmt.Println()

	// タスクサマリー
	fmt.Println("📋 Task Summary:")
	fmt.Printf("   Total Tasks: %d\n", report.TotalTasks)
	fmt.Printf("   ✅ Successful: %d\n", report.SuccessTasks)
	fmt.Printf("   ❌ Failed: %d\n", report.FailedTasks)
	fmt.Printf("   ⏭️  Skipped: %d\n", report.SkippedTasks)
	fmt.Println()

	// 各タスクの詳細
	fmt.Println("📝 Task Details:")
	for i, task := range report.TaskResults {
		fmt.Printf("\n%d. %s\n", i+1, task.TaskName)
		fmt.Printf("   Duration: %s\n", task.Duration.Round(time.Millisecond))
		fmt.Printf("   Files: %d total (%d success, %d failed, %d skipped)\n",
			task.TotalFiles, task.SuccessFiles, task.FailedFiles, task.SkippedFiles)
		fmt.Printf("   Total Size: %s\n", formatBytes(task.TotalBytes))

		if task.Error != nil {
			fmt.Printf("   ❌ Error: %v\n", task.Error)
		}

		// 失敗したファイルの詳細（最大5件）
		if task.FailedFiles > 0 {
			fmt.Println("   Failed Files:")
			count := 0
			for _, ur := range task.UploadResults {
				if !ur.Success && ur.Error != nil {
					fmt.Printf("     - %s: %v\n", ur.Source, ur.Error)
					count++
					if count >= 5 && task.FailedFiles > 5 {
						fmt.Printf("     ... and %d more\n", task.FailedFiles-5)
						break
					}
				}
			}
		}
	}

	fmt.Println("\n" + strings.Repeat("=", 60))
}

// formatBytes バイト数を人間が読みやすい形式に変換
func formatBytes(bytes int64) string {
	const (
		KB = 1024
		MB = KB * 1024
		GB = MB * 1024
	)

	switch {
	case bytes >= GB:
		return fmt.Sprintf("%.2f GB", float64(bytes)/GB)
	case bytes >= MB:
		return fmt.Sprintf("%.2f MB", float64(bytes)/MB)
	case bytes >= KB:
		return fmt.Sprintf("%.2f KB", float64(bytes)/KB)
	default:
		return fmt.Sprintf("%d B", bytes)
	}
}
